src/utils/fixedptc.h

```diff
@@ -194,66 +197,108 @@ static inline fixedpt fixedpt_div(fixedpt A, fixedpt B) {
+    char acc[FIXEDPT_WCHARS] = {0};
     fixedptud fr, ip;
+    fixedptud magnitude = (fixedptud)A;
     const fixedptud one = (fixedptud)1 << FIXEDPT_BITS;
     const fixedptud mask = one - 1;
+    size_t num_places = 0;
+
+    // return if the destination is null or too small
+    if(str == NULL || bufsize < 2) {
+        return 0;
+    }
 
-    if(max_dec == -1)
+    if(max_dec == -1) {
 #if FIXEDPT_BITS == 16 || FIXEDPT_BITS == 32
 #if FIXEDPT_FBITS < 16
-        max_dec = 2;
+        num_places = 2;
 #else
-        max_dec = 4;
+        num_places = 4;
 #endif
 #elif FIXEDPT_BITS == 64
-        max_dec = 10;
+        num_places = 10;
 #else
 #error Invalid width
 #endif
-    else if(max_dec == -2)
-        max_dec = 15;
+    } else if(max_dec == -2) {
+        num_places = 15;
+    } else if(max_dec == -3) {
+        num_places = 15;
+    } else if(max_dec < 0) {
+        // other negative values of max_dec not allowed
+        return 0;
+    } else {
+        num_places = max_dec;
+    }
 
     if(A < 0) {
         str[slen++] = '-';
-        A *= -1;
+        magnitude *= -1;
     }
 
-    ip = fixedpt_toint(A);
+    // Accumulate the integer digits in reverse order
+    // by dividing the integer component by 10 each time.
+    // Use do/while so at least a 0 is written
+    ip = fixedpt_toint(magnitude);
     do {
-        tmp[ndec++] = '0' + ip % 10;
+        acc[nint++] = '0' + ip % 10;
         ip /= 10;
-    } while(ip != 0);
+    } while(ip != 0 && nint < sizeof(acc) - 1);
 
-    while(ndec > 0)
-        str[slen++] = tmp[--ndec];
-    str[slen++] = '.';
+    // reverse the accumulated digits into the output string
+    while(nint > 0 && slen < bufsize - 1) {
+        str[slen++] = acc[--nint];
+    }
 
-    fr = (fixedpt_fracpart(A) << FIXEDPT_WBITS) & mask;
-    do {
-        fr = (fr & mask) * 10;
+    // check if we want, and have space for, decimal digits
+    if(num_places > 0 && slen < bufsize - 1) {
+        str[slen++] = '.';
 
-        str[slen++] = '0' + (fr >> FIXEDPT_BITS) % 10;
-        ndec++;
-    } while(fr != 0 && ndec < max_dec);
+        fr = (fixedpt_fracpart(magnitude) << FIXEDPT_WBITS) & mask;
+        while(nfrac < num_places && slen < bufsize - 1) {
+            fr = (fr & mask) * 10;
 
-    if(ndec > 1 && str[slen - 1] == '0')
-        str[slen - 1] = '\0'; /* cut off trailing 0 */
-    else
-        str[slen] = '\0';
+            str[slen++] = '0' + (fr >> FIXEDPT_BITS) % 10;
+            nfrac++;
+        }
+
+        // trim any trailing 0s, if desired
+        while(max_dec == -3 && nfrac > 0 && str[slen - 1] == '0') {
+            slen--;
+            nfrac--;
+        }
+
+        // if we didn't add any decimal values, omit the decimal place
+        if(nfrac == 0) {
+            slen--;
+        }
+    }
+
+    int bytes_used = slen < bufsize ? slen : bufsize - 1;
+    // always null terminate because we will always write at least one byte
+    str[bytes_used] = '\0';
```

Comment:
```suggestion
    str[bytes_used] = '\0';
```
I don't understand this comment, but I don't think the comment is needed.